### DISCUSSION TITLE
extra_query parameter in OpenAI.batches.list

### DISCUSSION THREAD

**Discussion by Cybernetic-Ransomware** [openai/openai-python#1675] [Q&A] [answered]
I am trying to filter batches on the server side using the extra_query parameter like this: 
`extra_query={'status': 'completed', 'created_before': '2024-08-26T00:00:00Z'}`

 Unfortunately, it's not working. I also couldn't find any examples. Is the extra_query parameter currently functional?

**Comment by tamish-max** [comment]
`extra_query` is a low-level escape hatch in the Python SDK — it appends raw query parameters to the HTTP request. It does **not** mean the backend supports those filters.

So if `batches.list` does not document `status` / `created_before` as valid server-side query params, `extra_query` will not magically enable them. The endpoint will either ignore them or reject them.

So the short answer is:
- yes, `extra_query` itself is functional;
- no, it only works for parameters the API endpoint actually understands.

For `batches.list`, I would assume client-side filtering unless/until those filters appear in the official API docs.

**Comment by Hikari-Tsai** [comment, marked as answer]
`extra_query` will append arbitrary query parameters to the HTTP request, but it does not make the server support filters that the endpoint does not define.

For `client.batches.list()`, the current API reference only documents `after` and `limit` as query parameters:
https://developers.openai.com/api/reference/resources/batches/methods/list/

So parameters like `status` or `created_before` are not server-side filters for this endpoint. You need to page through the batches and filter client-side, for example:

```python
from datetime import datetime, timezone

cutoff = datetime.fromisoformat("2024-08-26T00:00:00+00:00")

for batch in client.batches.list(limit=100):
    created_at = datetime.fromtimestamp(batch.created_at, tz=timezone.utc)
    if batch.status == "completed" and created_at < cutoff:
        print(batch.id)
```

If you have many batches, use pagination (`limit` / `after`) or the SDK's auto-pagination pattern and stop once your ordering assumptions make it safe.

**Reply by Cybernetic-Ransomware** [reply]
Thanks, that clarifies it.

So `extra_query` only appends arbitrary query parameters to the request, but `batches.list()` currently supports only `after` and `limit` server-side. Filtering by `status` or `created_at` has to be done client-side.

That works, but it means clients with many historical batches must paginate through potentially all batches just to find completed or older ones. Server-side filters for `status` and creation timestamps would be very useful here.
